### ISSUE TITLE
Test failure in add_sssaaaaaa

### ISSUE THREAD

**Issue by orlitzky** [flintlib/flint#2117]
While preparing https://github.com/flintlib/flint/pull/2116, I tried to run the tests from my git checkout, and got this failure:

```
$ make check
/usr/bin/mkdir -p build/test
  CC  test/main.c
add_ssaaaa...
add_ssaaaa                                        0.03   (PASS)
add_sssaaaaaa...
add_sssaaaaaa                                     0.00   (FAIL)
Aliasing: 1
n = 5
a = [3, 132912939, 22630, 15035479, 12062733498991]
b = [36930710888, 27284933481435, 387565, 65535, 37138525025]
s = [36930710891, 27285066394374, 410195, 30202028, 12099872024016]
t = [36930710891, 27285066394374, 410195, 15101014, 12099872024016]
make: *** [Makefile:771: build/test/main_TEST_RUN] Aborted
```

**Comment by albinahlback** [comment]
What CPU, OS and compiler do you use? What settings did you push to Autotools?

**Comment by orlitzky** [comment]
This is a 64-bit RISC-V machine running Gentoo linux. I tried with both GCC 14.2.1 and GCC 15.0.0 20241117 (a snapshot, it isn't released yet). Basic steps were `autoreconf -fi; configure; make; make check` although I tried with our system package as well which uses CMake and forces NTL support to be enabled.

In all cases, the latest release (with https://github.com/flintlib/flint/pull/2116 applied) works, but git HEAD has this failure.

**Comment by albinahlback** [comment]
Thanks for this information. Can you make sure that it the definition of `add_s..a...` that appears in `longlong.h` that is being used? Change the definition to something like
```c
# define add_sssaaaaaa(s2, s1, s0, a2, a1, a0, b2, b1, b0) \
    hocus pocus
```
and verify that it generates an error when trying to compile. It should be at line 118 in `longlong.h`.

**Comment by albinahlback** [comment]
Hmm, I thought I fixed this in `e5e70d8b36d02a25173a98bb5339e8e4739bd214`. I suppose this arose as of GCC 14, as I recall testing with the latest release of GCC on Arch Linux. I don't know how to fix this for generic compilers. For GCC compilers, of course, the easiest is to use compiler builtins like `__builtin_adc` or whatever.

**Comment by orlitzky** [comment]
If it's useful, here are the results for `n=6` and `n=7`:

```
add_sssaaaaaa                                     0.00   (FAIL)
Aliasing: 1
n = 6
a = [3, 132912939, 22630, 15035479, 12062733498991, 256]
b = [36930710888, 27284933481435, 387565, 65535, 37138525025, 16384]
s = [36930710891, 27285066394374, 410195, 30202028, 12099872024016, 16640]
t = [36930710891, 27285066394374, 410195, 15101014, 12099872024016, 16640]
```

```
add_sssaaaaaa                                     0.00   (FAIL)
Aliasing: 1
n = 7
a = [3, 132912939, 22630, 15035479, 12062733498991, 256, 4310123375852123]
b = [36930710888, 27284933481435, 387565, 65535, 37138525025, 16384, 131072]
s = [36930710891, 27285066394374, 410195, 30202028, 12099872024016, 16640, 4310123375983195]
t = [36930710891, 27285066394374, 410195, 15101014, 12099872024016, 16640, 4310123375983195]
```

**Comment by orlitzky** [comment]
Interestingly enough, with clang-19.1.3,

```
$ make check
/usr/bin/mkdir -p build/test
  CC  test/main.c
add_ssaaaa...
add_ssaaaa                                        0.02   (PASS)
add_sssaaaaaa...
add_sssaaaaaa                                     0.14   (PASS)
flint_clz...
flint_clz                                         0.00   (PASS)
flint_ctz...
flint_ctz                                         0.00   (PASS)
flint_fprintf...
flint_fprintf                                     0.03   (PASS)
flint_printf...
flint_printf                                          (SKIPPED)
memory_manager...
memory_manager                                    0.02   (PASS)
sdiv_qrnnd...
sdiv_qrnnd                                        0.00   (FAIL)
Check n = q * d + r failed
nh = 10004721567595360, nl = 81
d = -9223372036854775808
q = -20009443135190882
r = 80
ph = 10004721567595441, pl = 80
```

**Comment by albinahlback** [comment]
Yes, but that uses Clang instrinsics.  See `longlong_asm_clang.h`, I believe it is called.

**Comment by albinahlback** [comment]
> Interestingly enough, with clang-19.1.3,
> 
> ```
> $ make check
> /usr/bin/mkdir -p build/test
>   CC  test/main.c
> add_ssaaaa...
> add_ssaaaa                                        0.02   (PASS)
> add_sssaaaaaa...
> add_sssaaaaaa                                     0.14   (PASS)
> flint_clz...
> flint_clz                                         0.00   (PASS)
> flint_ctz...
> flint_ctz                                         0.00   (PASS)
> flint_fprintf...
> flint_fprintf                                     0.03   (PASS)
> flint_printf...
> flint_printf                                          (SKIPPED)
> memory_manager...
> memory_manager                                    0.02   (PASS)
> sdiv_qrnnd...
> sdiv_qrnnd                                        0.00   (FAIL)
> Check n = q * d + r failed
> nh = 10004721567595360, nl = 81
> d = -9223372036854775808
> q = -20009443135190882
> r = 80
> ph = 10004721567595441, pl = 80
> ```

Sorry, is this still on RISC?

**Comment by orlitzky** [comment]
> Sorry, is this still on RISC?

Yep.

**Comment by albinahlback** [comment]
Solved as of #2122. Please reopen if the issue has not been solved.

**Comment by orlitzky** [comment]
Looks good with GCC 15 and I'll try with clang, too. Thanks, I applaud your dedication.
